File: src-tauri/plugins/tauri-plugin-hbut-background/android/src/main/kotlin/com/hbut/mini/background/BackgroundStore.kt
// Kotlin 端持久化：与 Rust store.rs 语义对齐。
// - 目录构造（dir: File）与 Context 解耦：JVM 单测用临时目录，真机用 Context.filesDir；
// - 原子写：先写 .tmp-* 再 rename；
// - 损坏/版本不兼容：备份为 .corrupt-* 后降级，不 crash；
// - 容量上限：EVENT_INBOX_CAP，超出丢弃最旧；
// - scope 清理：账号切换/退出时完整清理 context/state/events。

package com.hbut.mini.background

import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/** 存储错误。 */
class StoreException(message: String, cause: Throwable? = null) : Exception(message, cause)

/** 插件数据目录内的文件名（与 Rust store.rs 一致，跨端契约的一部分）。 */
object StoreFiles {
    const val CONFIG = "config.json"
    const val CONTEXT = "context.json"
    const val STATE = "state.json"
    const val EVENTS = "events.json"
}

/** 落盘存储实现。 */
class BackgroundStore(private val dir: File) {

    init {
        if (!dir.exists() && !dir.mkdirs()) {
            throw StoreException("创建存储目录失败: $dir")
        }
    }

    fun dir(): File = dir

    private fun path(name: String): File = File(dir, name)

    // ---- config ----

    fun loadConfig(): BackgroundConfig {
        val obj = loadChecked<JSONObject>(StoreFiles.CONFIG) ?: return BackgroundConfig(
            schema = BG_SCHEMA_VERSION,
            enabled = false,
            intervalMinutes = 30,
            business = emptyList(),
            scope = null,
        )
        return BackgroundConfig.fromJson(obj)
    }

    fun saveConfig(config: BackgroundConfig) {
        saveAtomic(StoreFiles.CONFIG, config.toJson())
    }

    // ---- context ----

    fun loadContext(): BackgroundContext? {
        val obj = loadChecked<JSONObject>(StoreFiles.CONTEXT) ?: return null
        return BackgroundContext.fromJson(obj)
    }

    fun saveContext(context: BackgroundContext) {
        saveAtomic(StoreFiles.CONTEXT, context.toJson())
    }

    // ---- state ----

    fun loadState(): BackgroundCheckState? {
        val obj = loadChecked<JSONObject>(StoreFiles.STATE) ?: return null
        return BackgroundCheckState.fromJson(obj)
    }

    fun saveState(state: BackgroundCheckState) {
        saveAtomic(StoreFiles.STATE, state.toJson())
    }

    // ---- events ----

    fun loadEvents(): List<BackgroundEvent> {
        val arr = loadChecked<JSONArray>(StoreFiles.EVENTS) ?: return emptyList()
        return (0 until arr.length()).map { BackgroundEvent.fromJson(arr.getJSONObject(it)) }
    }

    fun saveEvents(events: List<BackgroundEvent>) {
        val kept = if (events.size > EVENT_INBOX_CAP) {
            events.subList(events.size - EVENT_INBOX_CAP, events.size)
        } else {
            events
        }
        saveAtomic(StoreFiles.EVENTS, JSONArray(kept.map { it.toJson() }))
    }

    fun appendEvent(event: BackgroundEvent) {
        saveEvents(loadEvents() + event)
    }

    /** 消费并移除事件；limit 为 null 时消费全部。 */
    fun consumeEvents(limit: Int?): ConsumeEventsResult {
        val events = loadEvents()
        val take = (limit ?: events.size).coerceAtMost(events.size)
        val consumed = events.subList(0, take)
        saveEvents(events.subList(take, events.size))
        return ConsumeEventsResult(BG_SCHEMA_VERSION, consumed, events.size - take)
    }

    // ---- scope 清理（账号切换/退出） ----

    /** 返回 (是否清除 context/state, 清除的事件数)。 */
    fun clearScope(scope: String): Pair<Boolean, Int> {
        var clearedAny = false
        loadContext()?.let { ctx ->
            if (ctx.scope == scope) {
                path(StoreFiles.CONTEXT).delete()
                clearedAny = true
            }
        }
        loadState()?.let { state ->
            if (state.scope == scope) {
                path(StoreFiles.STATE).delete()
                clearedAny = true
            }
        }
        val events = loadEvents()
        val kept = events.filter { it.scope != scope }
        saveEvents(kept)
        val removed = events.size - kept.size
        if (removed > 0) clearedAny = true
        return clearedAny to removed
    }

    // ---- 内部实现 ----

    /** 读取 + schema 版本校验；损坏/版本不兼容时备份后返回 null（安全降级）。 */
    private fun <T> loadChecked(name: String): T? {
        val file = path(name)
        if (!file.exists()) return null
        val text = try {
            file.readText()
        } catch (e: Exception) {
            return null
        }
        return try {
            @Suppress("UNCHECKED_CAST")
            val parsed = parseJson(text) as T
            val schema = (parsed as? JSONObject)?.optInt("schema")
                ?: (parsed as? JSONArray)?.let { arr ->
                    if (arr.length() > 0) arr.getJSONObject(0).optInt("schema") else BG_SCHEMA_VERSION
                }
            if (schema == BG_SCHEMA_VERSION) {
                parsed
            } else {
                backupCorrupt(name, file)
                null
            }
        } catch (e: Exception) {
            backupCorrupt(name, file)
            null
        }
    }

    private fun parseJson(text: String): Any {
        val trimmed = text.trim()
        return if (trimmed.startsWith("[")) JSONArray(trimmed) else JSONObject(trimmed)
    }

    /** 损坏/版本不兼容文件备份为 *.corrupt-<ts>。 */
    private fun backupCorrupt(name: String, file: File) {
        try {
            val backup = File(dir, "$name.corrupt-${System.currentTimeMillis()}")
            file.renameTo(backup)
        } catch (_: Exception) {
            // 备份失败不阻断降级路径
        }
    }

    /** 原子写：tmp 文件 + rename。 */
    private fun saveAtomic(name: String, value: Any) {
        val tmp = File(dir, "$name.tmp-${System.currentTimeMillis()}")
        tmp.writeText(value.toString())
        val target = path(name)
        if (!tmp.renameTo(target)) {
            tmp.delete()
            throw StoreException("原子写入失败: $name")
        }
    }
}

File: src-tauri/plugins/tauri-plugin-hbut-background/ios/Sources/HbutBackgroundPlugin/BackgroundStore.swift
// Swift 端持久化：与 Rust store.rs / Kotlin BackgroundStore.kt 语义对齐。
// - 原子写：Data.write(options: .atomic)；
// - 损坏/版本不兼容：备份为 *.corrupt-<ts> 后降级，不 crash；
// - 容量上限：eventInboxCap，超出丢弃最旧；
// - scope 清理：账号切换/退出时完整清理 context/state/events。

import Foundation

/// 存储错误。
public enum StoreError: Error, LocalizedError {
    case dirInit(String)
    case read(String)
    case write(String)

    public var errorDescription: String? {
        switch self {
        case .dirInit(let m), .read(let m), .write(let m): return m
        }
    }
}

/// 落盘存储：目录构造与 App 容器解耦（真机由调用方传 applicationSupportURL 等目录）。
public class BackgroundStore {
    public static let configFile = "config.json"
    public static let contextFile = "context.json"
    public static let stateFile = "state.json"
    public static let eventsFile = "events.json"

    private let dir: URL
    private let fileManager: FileManager

    public init(dir: URL) throws {
        self.dir = dir
        self.fileManager = FileManager.default
        try fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
    }

    public func directory() -> URL { dir }

    private func path(_ name: String) -> URL { dir.appendingPathComponent(name) }

    // ---- config ----

    public func loadConfig() -> BackgroundConfig {
        guard let data = loadChecked(name: Self.configFile) else {
            return BackgroundConfig()
        }
        return (try? JSONDecoder().decode(BackgroundConfig.self, from: data)) ?? BackgroundConfig()
    }

    public func saveConfig(_ config: BackgroundConfig) throws {
        try saveAtomic(name: Self.configFile, value: config)
    }

    // ---- context ----

    public func loadContext() -> BackgroundContext? {
        guard let data = loadChecked(name: Self.contextFile) else { return nil }
        return try? JSONDecoder().decode(BackgroundContext.self, from: data)
    }

    public func saveContext(_ context: BackgroundContext) throws {
        try saveAtomic(name: Self.contextFile, value: context)
    }

    // ---- state ----

    public func loadState() -> BackgroundCheckState? {
        guard let data = loadChecked(name: Self.stateFile) else { return nil }
        return try? JSONDecoder().decode(BackgroundCheckState.self, from: data)
    }

    public func saveState(_ state: BackgroundCheckState) throws {
        try saveAtomic(name: Self.stateFile, value: state)
    }

    // ---- events ----

    public func loadEvents() -> [BackgroundEvent] {
        guard let data = loadChecked(name: Self.eventsFile) else { return [] }
        return (try? JSONDecoder().decode([BackgroundEvent].self, from: data)) ?? []
    }

    public func saveEvents(_ events: [BackgroundEvent]) throws {
        let kept = events.count > eventInboxCap
            ? Array(events.suffix(eventInboxCap))
            : events
        try saveAtomic(name: Self.eventsFile, value: kept)
    }

    public func appendEvent(_ event: BackgroundEvent) throws {
        try saveEvents(loadEvents() + [event])
    }

    /// 消费并移除事件；limit 为 nil 时消费全部。
    public func consumeEvents(limit: Int?) throws -> ConsumeEventsResult {
        let events = loadEvents()
        let take = min(limit ?? events.count, events.count)
        let consumed = Array(events.prefix(take))
        try saveEvents(Array(events.dropFirst(take)))
        return ConsumeEventsResult(events: consumed, remaining: events.count - take)
    }

    // ---- scope 清理 ----

    /// 返回 (是否清除 context/state, 清除的事件数)。
    @discardableResult
    public func clearScope(_ scope: String) throws -> (cleared: Bool, removedEvents: Int) {
        var clearedAny = false
        if let ctx = loadContext(), ctx.scope == scope {
            try? fileManager.removeItem(at: path(Self.contextFile))
            clearedAny = true
        }
        if let state = loadState(), state.scope == scope {
            try? fileManager.removeItem(at: path(Self.stateFile))
            clearedAny = true
        }
        let events = loadEvents()
        let kept = events.filter { $0.scope != scope }
        try saveEvents(kept)
        let removed = events.count - kept.count
        if removed > 0 { clearedAny = true }
        return (clearedAny, removed)
    }

    // ---- 内部实现 ----

    /// 读取 + schema 版本校验；损坏/版本不兼容时备份后返回 nil（安全降级）。
    private func loadChecked(name: String) -> Data? {
        let url = path(name)
        guard fileManager.fileExists(atPath: url.path) else { return nil }
        guard let data = try? Data(contentsOf: url) else { return nil }
        let schema = extractSchema(data: data)
        if schema == nil || schema != bgSchemaVersion {
            backupCorrupt(name: name, url: url)
            return nil
        }
        return data
    }

    /// 从数据中提取顶层 schema 字段（对象或数组首元素）。
    private func extractSchema(data: Data) -> Int? {
        guard let json = try? JSONSerialization.jsonObject(with: data) else { return nil }
        if let obj = json as? [String: Any], let schema = obj["schema"] as? Int {
            return schema
        }
        if let arr = json as? [[String: Any]], let first = arr.first {
            return first["schema"] as? Int
        }
        return nil
    }

    private func backupCorrupt(name: String, url: URL) {
        let backup = dir.appendingPathComponent("\(name).corrupt-\(Int(Date().timeIntervalSince1970 * 1000))")
        try? fileManager.moveItem(at: url, to: backup)
    }

    /// 原子写（Data.write .atomic 自带 tmp+rename 语义）。
    private func saveAtomic(name: String, value: Encodable) throws {
        do {
            let data = try JSONEncoder().encode(value)
            try data.write(to: path(name), options: .atomic)
        } catch {
            throw StoreError.write("写入 \(name) 失败: \(error)")
        }
    }
}
